### ISSUE TITLE
Possible (virtual users and filters) bug using LOGIN_NAME_MAX define

### ISSUE THREAD

**Issue by terminar** [OpenSMTPD/OpenSMTPD#1020]
As mentioned at https://github.com/poolpOrg/poolp.org_posts/issues/31 i had problems on FreeBSD 12.1 with the filters not containing the full username which a user provides in the filter event "link-auth".

I tried to debug that with 6.6.0p1 and current github master. It seems that the use of the define **LOGIN_NAME_MAX** is maybe critical on non-OpenBSD systems for virtual users. OpenBSD seems to use 32 chars (https://github.com/openbsd/src/blob/master/sys/sys/syslimits.h). When OpenSMTPD is compiled on FreeBSD it seems to use the openbsd-compat/defines.h from OpenSMTPD itself which sets
```
#ifndef LOGIN_NAME_MAX
# define LOGIN_NAME_MAX 9
#endif
``` 
on my system.
That results to a (maybe problematic) situation e.g. at https://github.com/OpenSMTPD/OpenSMTPD/blob/f15f16c8e5de7ce271994377832ea325d0e8d1f5/smtpd/smtp_session.c#L971

Given that ```char				 user[LOGIN_NAME_MAX];``` will contain ```user[9]``` the call to 
```
strnvis(user, s->username, sizeof user, VIS_WHITE | VIS_SAFE);
```
will cut a username like "terminar@example.org" to "terminar" and remove the whole "virtual users domain part" / cut it at the 9th position due to the ```sizeof user``` call.

I found references to LOGIN_NAME_MAX at mda.c, smtp_session.c and smtpd.c but can't evaluate if this is just cosmetical or will affect authentication of virtual users (and even if my assumtion is correct).

Maybe someone can take a look at it?

**Comment by poolpOrg** [comment]
Hello,

Good catch !

First, the openbsd-compat LOGIN_NAME_MAX should be aligned with OpenBSD's, I don't know where that 9 came from but it's not right.

Then, the user buffer in smtp_session should not be LOGIN_NAME_MAX but probably SMTPD_MAXMAILADDRSIZE. I'll commit the fix tomorrow as my mind is blurry now.

Don't worry too much about the truncation, the user buffer is only used for logging and reporting _upon return_ of the actual authentication which takes place using the correct buffer. Basically what happens is that when we're back from authentication and have a result, we want to log the username and to ensure that there are no strange characters ending in a logfile, we use strnvis() to sanitize the buffer before it hits a file, there's no logic that depends on that buffer beyond logging/reporting.

**Comment by poolpOrg** [comment]
Hello,

I have committed a fix upstream, it's been pushed to github to both master and portable branch.

As explained in previous comment, the real fix was bumping the user buffer from LOGIN_NAME_MAX to SMTPD_MAXMAILADDRSIZE, but I also aligned LOGIN_NAME_MAX to the value used in OpenBSD. After investigation, it was set to 9 because that's the POSIX value, most systems no longer limit usernames to 8 chars + nul-byte so 32 seems like a correct value. There are only two places which use LOGIN_NAME_MAX, I might even manage to get rid of this and remove the define from compat layer later.
